pkg/kubelet/cm/deviceplugin/pod_devices.go

```diff
@@ -252,6 +255,20 @@ func (pdev podDevices) deviceRunContainerOptions(podUID, contName string) *Devic
 				SELinuxRelabel: false,
 			})
 		}
+
+		// Updates for Annotations
+		for k, v := range resp.Annotations {
+			if e, ok := annotationsMap[k]; ok {
+				glog.V(4).Infof("skip existing annotation %s %s", k, v)
+				if e != v {
+					glog.Errorf("Annotation %s has conflicting setting: %s and %s", k, e, v)
+				}
```

Comment:
Should we concatenate values instead of just taking the first one, especially given that the immediate use case you have in mind is to support runc prestart hook? Suppose two kinds of devices both want to follow this approach of running their own device-specific runc hook, and a container requires to use both resources, the behavior is undefined if we just take the first one. In that case, specifying runc hook information in pod spec annotations perhaps have better defined behavior, which I think is the approach we should encourage people to follow.

In general, I agree with @Random-Liu that it's good to enable people to experiment new ideas via Annotations. On the other hand, I am also concerned of using annotations to pass per-device-type runc hook information which could lead to undefined behavior. If we decide to extend the device plugin API to include container Annotations, we should carefully document what are the recommended uses and the behavior when we have conflicting annotation settings.

Reply:
> Should we concatenate values instead of just taking the first one, especially given that the immediate use case you have in mind is to support runc prestart hook? Suppose two kinds of devices both want to follow this approach of running their own device-specific runc hook, and a container requires to use both resources, the behavior is undefined if we just take the first one. In that case, specifying runc hook information in pod spec annotations perhaps have better defined behavior, which I think is the approach we should encourage people to follow.

For the merging part isn't it the same argument as for environment variable ? It seems sane to have.

> especially given that the immediate use case you have in mind is to support runc prestart hook.

CRIO [hooks](https://github.com/kubernetes-incubator/cri-o/blob/master/hooks.md) works by registering the path to the hook and the annotation that needs to be present on the container for the hook to be invoked.
The case you are mentioning is not a problem if every device plugin uses a different hook.

In general only the nvidia device plugin is expected to use Annotations. And device plugins are expected to be thoroughly vetted by the community.

> On the other hand, I am also concerned of using annotations to pass per-device-type runc hook information which could lead to undefined behavior.

This is the only officially NVIDIA supported way to run GPU enabled container. Not using it will lead to you not being able to run features we add (Graphics, MPS, MPI, ...).

> In that case, specifying runc hook information in pod spec annotations perhaps have better defined behavior, which I think is the approach we should encourage people to follow.

In general the plan is on the long term to standardize this in CRI through an explicit API.
I don't know if we want to expose this to a user though.

Reply:
What are the semantics CRIO expects for specifying runc hook in container Annotations? What is the key and what is the value?

I also don't agree with the argument that "In general only the nvidia device plugin is expected to use Annotations". Once we add it in the API, how can you prevent other vendors from using it?

Reply:
> What are the semantics CRIO expects for specifying runc hook in container Annotations? What is the key and what is the value?

See the doc I linked in the previous comment explaining the CRIO hooks.

> I also don't agree with the argument that "In general only the nvidia device plugin is expected to use Annotations". Once we add it in the API, how can you prevent other vendors from using it?

Quoting @vishh on this: "device plugins are special, where they are expected to be closely audited"
I can understand the argument though but as mentioned higher by you, "it's good to enable people to experiment new ideas via Annotations" which btw the device plugin is the only place we can add them to a container.

Reply:
Looking at the crio doc, I now understand its usage of container Annotations is specific to hook config. I think in this case, we should explicitly mention in the device plugin API comment and also the device plugin user guide that keys in Annotations should be specific to the device plugin, and if it conflicts with any other device plugin, behavior is un-defined. Also agree that we should document the same for env variables.

I still don't agree that device plugin is the only place to add Annotations to a container but just in terms of device plugin API, I am fine with adding this extension. I will leave the final decision to @dchen1107 and @derekwaynecarr for whether we want to take container annotations from device plugins.

Reply:
> I still don't agree that device plugin is the only place to add Annotations

Where would you add them ?

Reply:
+1 for adding the documentation about the requirement that keys should be unique.
Also I would like to see this extension added to device plugins. This will provide a generic mechanism for invoking hooks, at least for now only with crio. But since crio is for k8s only that further favors addition of this extension.